File: framework/dotnet/BetterPortal/Context.cs
using System.Collections.Frozen;
using System.Text.RegularExpressions;

namespace BetterPortal;

public static class HttpAddress
{
    public static string Origin(string value, bool allowPath = false)
    {
        if (value.Length > 8192 || value.Any(c => char.IsWhiteSpace(c) || c < 32 || c == 127) || value.Contains('\\'))
            throw new ArgumentException("Invalid HTTP address");
        var address = Regex.Match(value, @"\A[^:]+://([^/?#]*)");
        var authority = address.Groups[1].Value;
        if (authority.Length == 0 || authority.Contains('@') || authority.Contains('%')
            || !Uri.TryCreate(value, UriKind.Absolute, out var uri) || uri.Scheme is not ("http" or "https") || uri.Host.Length == 0 || uri.Port == 0)
            throw new ArgumentException("Invalid HTTP authority");
        if (!allowPath && value[address.Length..] is not ("" or "/"))
            throw new ArgumentException("Expected an origin without path, query or fragment");
        var host = uri.IdnHost.ToLowerInvariant();
        if (host.Contains(':')) host = "[" + System.Net.IPAddress.Parse(host.Trim('[', ']')).ToString() + "]";
        return uri.Scheme + "://" + host + (uri.IsDefaultPort ? "" : ":" + uri.Port);
    }
    internal static IEnumerable<string> Origins(string hostname) => hostname.Contains("://", StringComparison.Ordinal)
        ? [Origin(hostname)] : [Origin("https://" + hostname), Origin("http://" + hostname)];
    internal static string Referer(string value)
    {
        var origin = Origin(value, allowPath: true);
        if (value.Contains('#')) throw new ArgumentException("Referer must not contain a fragment");
        var path = new Uri(value).PathAndQuery;
        return origin + (path == "/" ? "" : path);
    }
}

public sealed record OriginPolicy(FrozenSet<string> Origins, FrozenSet<string> Referers)
{
    public static OriginPolicy FromApp(Dictionary<string, object?> app)
    {
        var origins = ((List<object?>)app["hostnames"]!).Cast<string>().SelectMany(HttpAddress.Origins)
            .Concat(((List<object?>)app["originOverrides"]!).Cast<string>().Select(value => HttpAddress.Origin(value))).ToFrozenSet(StringComparer.Ordinal);
        return new(origins, origins.Concat(((List<object?>)app["refererOverrides"]!).Cast<string>().Select(HttpAddress.Referer)).ToFrozenSet(StringComparer.Ordinal));
    }
    public bool Allows(string? value, bool referer = false)
    {
        if (value is null) return false;
        try
        {
            return referer ? Referers.Contains(HttpAddress.Referer(value)) || Referers.Contains(HttpAddress.Origin(value, allowPath: true))
                : Origins.Contains(HttpAddress.Origin(value));
        }
        catch (ArgumentException) { return false; }
        catch (UriFormatException) { return false; }
    }
}

public sealed record ScopedContext(Dictionary<string, object?> Tenant, Dictionary<string, object?> App, OriginPolicy OriginPolicy)
{
    public string TenantId => (string)Tenant["id"]!;
    public string AppId => (string)App["id"]!;
}

/// <summary>One validated scoped snapshot. Request contexts receive owned copies.</summary>
public sealed class ScopedConfig
{
    private readonly Dictionary<string, object?> snapshot;
    private readonly Dictionary<string, Dictionary<string, object?>> tenants;
    private readonly Dictionary<string, Dictionary<string, object?>> apps;
    private readonly Dictionary<string, string> hosts = new(StringComparer.Ordinal);
    private readonly Dictionary<string, OriginPolicy> policies = new(StringComparer.Ordinal);
    public ScopedConfig(object value)
    {
        snapshot = (Dictionary<string, object?>)Contracts.Parse("ScopedServiceConfigSchema", value)!;
        tenants = Index((List<object?>)snapshot["tenants"]!);
        apps = Index((List<object?>)snapshot["apps"]!);
        foreach (var tenant in tenants.Values) Index((List<object?>)tenant["services"]!);
        foreach (var (id, app) in apps)
        {
            if (!tenants.ContainsKey((string)app["tenantId"]!)) throw new ArgumentException("App refers to an unknown tenant");
            policies[id] = OriginPolicy.FromApp(app);
            foreach (var origin in ((List<object?>)app["hostnames"]!).Cast<string>().SelectMany(HttpAddress.Origins))
            {
                if (hosts.TryGetValue(origin, out var existing) && existing != id) throw new ArgumentException("App host is ambiguous");
                hosts[origin] = id;
            }
        }
        foreach (var app in Index((List<object?>?)snapshot.GetValueOrDefault("configApps") ?? []).Values)
            if (!tenants.ContainsKey((string)app["tenantId"]!)) throw new ArgumentException("Config app refers to an unknown tenant");
        foreach (var origin in ((List<object?>)snapshot["managementOrigins"]!).Cast<string>()) HttpAddress.Origin(origin);
    }
    private static Dictionary<string, Dictionary<string, object?>> Index(IEnumerable<object?> values)
    {
        var result = new Dictionary<string, Dictionary<string, object?>>(StringComparer.Ordinal);
        foreach (var value in values.Cast<Dictionary<string, object?>>())
            if (!result.TryAdd((string)value["id"]!, value)) throw new ArgumentException("Duplicate scoped identity");
        return result;
    }
    private static Dictionary<string, object?> Copy(Dictionary<string, object?> value) => (Dictionary<string, object?>)Json.Read(Json.Write(value))!;
    public Dictionary<string, object?> Document() => Copy(snapshot);
    public ScopedContext? ById(string tenantId, string appId)
    {
        if (!apps.TryGetValue(appId, out var app) || !tenants.TryGetValue(tenantId, out var tenant) || tenant["active"] is not true
            || !Equals(app["tenantId"], tenantId)) return null;
        return new(Copy(tenant), Copy(app), policies[appId]);
    }
    /// <summary>Only host-verified proxy addresses are accepted; raw proxy/HTMX/scope headers confer no authority.</summary>
    public ScopedContext? Resolve(IReadOnlyDictionary<string, string> headers, string scheme = "https", string mode = "service", IEnumerable<string>? trustedAddresses = null)
    {
        if (scheme is not ("http" or "https") || mode is not ("service" or "theme")) throw new ArgumentException("Invalid request addressing mode");
        var normalized = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var (key, value) in headers) if (!normalized.TryAdd(key, value)) throw new ArgumentException("Duplicate request headers");
        var direct = new[] { "host", ":authority", "authority" }.Where(normalized.ContainsKey).Select(key => (Url: scheme + "://" + normalized[key], AllowPath: false));
        var embedded = new[] { "origin", "referer" }.Where(normalized.ContainsKey).Select(key => (Url: normalized[key], AllowPath: key == "referer"));
        var trusted = (trustedAddresses ?? []).Select(value => (Url: value, AllowPath: false));
        var candidates = mode == "theme" ? direct.Concat(trusted).Concat(embedded) : embedded.Concat(direct).Concat(trusted);
        if (normalized.TryGetValue("alt-used", out var alternative)) candidates = candidates.Append((scheme + "://" + alternative, false));
        foreach (var candidate in candidates)
        {
            string origin;
            try { origin = HttpAddress.Origin(candidate.Url, candidate.AllowPath); }
            catch (ArgumentException) { continue; }
            catch (UriFormatException) { continue; }
            if (hosts.TryGetValue(origin, out var id) && ById((string)apps[id]["tenantId"]!, id) is { } context) return context;
        }
        return null;
    }
}

File: framework/python/betterportal/context.py
"""Validated scoped configuration and exact browser addressing, independent of ASGI."""
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass
import ipaddress
from typing import Any, Mapping, Iterable
from urllib.parse import urlsplit

import httpx
from .contracts import parse


def http_origin(value: str, *, allow_path: bool = False) -> str:
    if not isinstance(value, str) or len(value) > 8192 or any(c.isspace() or ord(c) < 32 or ord(c) == 127 for c in value) or "\\" in value:
        raise ValueError("Invalid HTTP address")
    raw = urlsplit(value)
    if raw.scheme not in ("http", "https") or not raw.netloc or "@" in raw.netloc or "%" in raw.netloc:
        raise ValueError("Invalid HTTP authority")
    if not allow_path and (raw.path not in ("", "/") or "?" in value or "#" in value):
        raise ValueError("Expected an origin without path, query or fragment")
    try:
        url = httpx.URL(value)
        host = url.raw_host.decode("ascii")
    except (httpx.InvalidURL, UnicodeError) as error:
        raise ValueError("Invalid HTTP address") from error
    if not host or url.port == 0:
        raise ValueError("Invalid HTTP authority")
    if ":" in host:
        host = "[" + ipaddress.IPv6Address(host).compressed + "]"
    port = url.port
    if port == (443 if url.scheme == "https" else 80):
        port = None
    return f"{url.scheme}://{host}" + (f":{port}" if port is not None else "")


def _origins(hostname: str) -> list[str]:
    return [http_origin(hostname)] if "://" in hostname else [http_origin("https://" + hostname), http_origin("http://" + hostname)]


def _referer(value: str) -> str:
    origin = http_origin(value, allow_path=True)
    if "#" in value:
        raise ValueError("Referer must not contain a fragment")
    path = httpx.URL(value).raw_path.decode("ascii")
    return origin + (path if path != "/" else "")


@dataclass(frozen=True)
class OriginPolicy:
    origins: frozenset[str]
    referers: frozenset[str]

    @classmethod
    def from_app(cls, app: Mapping[str, Any]) -> OriginPolicy:
        origins = {origin for hostname in app["hostnames"] for origin in _origins(hostname)}
        origins.update(http_origin(value) for value in app["originOverrides"])
        return cls(frozenset(origins), frozenset(origins | {_referer(value) for value in app["refererOverrides"]}))

    def allows(self, value: str | None, *, referer: bool = False) -> bool:
        if value is None:
            return False
        try:
            if referer:
                normalized = _referer(value)
                return normalized in self.referers or http_origin(value, allow_path=True) in self.referers
            return http_origin(value) in self.origins
        except ValueError:
            return False


@dataclass(frozen=True)
class ScopedContext:
    tenant: dict[str, Any]
    app: dict[str, Any]
    origin_policy: OriginPolicy

    @property
    def tenant_id(self) -> str:
        return self.tenant["id"]

    @property
    def app_id(self) -> str:
        return self.app["id"]


class ScopedConfig:
    """One validated snapshot. Request contexts own copies; replacement belongs to the provider."""
    def __init__(self, value: Mapping[str, Any]):
        self._snapshot = parse("ScopedServiceConfigSchema", deepcopy(dict(value)))
        self._tenants = self._index(self._snapshot["tenants"])
        self._apps = self._index(self._snapshot["apps"])
        self._hosts: dict[str, str] = {}
        self._policies = {}
        for tenant in self._tenants.values():
            self._index(tenant["services"])
        for app in self._apps.values():
            if app["tenantId"] not in self._tenants:
                raise ValueError("App refers to an unknown tenant")
            self._policies[app["id"]] = OriginPolicy.from_app(app)
            for hostname in app["hostnames"]:
                for origin in _origins(hostname):
                    if origin in self._hosts and self._hosts[origin] != app["id"]:
                        raise ValueError("App host is ambiguous")
                    self._hosts[origin] = app["id"]
        for app in self._snapshot.get("configApps", []):
            if app["tenantId"] not in self._tenants:
                raise ValueError("Config app refers to an unknown tenant")
        self._index(self._snapshot.get("configApps", []))
        for origin in self._snapshot["managementOrigins"]:
            http_origin(origin)

    @staticmethod
    def _index(values: Iterable[dict[str, Any]]) -> dict[str, dict[str, Any]]:
        result = {}
        for value in values:
            if value["id"] in result:
                raise ValueError("Duplicate scoped identity")
            result[value["id"]] = value
        return result

    def document(self) -> dict[str, Any]:
        return deepcopy(self._snapshot)

    def by_id(self, tenant_id: str, app_id: str) -> ScopedContext | None:
        app, tenant = self._apps.get(app_id), self._tenants.get(tenant_id)
        if app is None or tenant is None or not tenant["active"] or app["tenantId"] != tenant_id:
            return None
        return ScopedContext(deepcopy(tenant), deepcopy(app), self._policies[app_id])

    def resolve(self, headers: Mapping[str, str], *, scheme: str = "https", mode: str = "service",
                trusted_addresses: Iterable[str] = ()) -> ScopedContext | None:
        """Proxy addresses must already be verified by the host's proxy middleware.

        Raw proxy, tenant/app and HTMX headers never establish authority here.
        """
        if scheme not in ("http", "https") or mode not in ("service", "theme"):
            raise ValueError("Invalid request addressing mode")
        normalized = {key.lower(): value for key, value in headers.items()}
        if len(normalized) != len(headers):
            raise ValueError("Duplicate request headers")
        direct = [(scheme + "://" + normalized[key], False) for key in ("host", ":authority", "authority") if key in normalized]
        embedded = [(normalized[key], key == "referer") for key in ("origin", "referer") if key in normalized]
        trusted = [(value, False) for value in trusted_addresses]
        candidates = direct + trusted + embedded if mode == "theme" else embedded + direct + trusted
        if "alt-used" in normalized:
            candidates.append((scheme + "://" + normalized["alt-used"], False))
        for candidate, allow_path in candidates:
            try:
                origin = http_origin(candidate, allow_path=allow_path)
            except ValueError:
                continue
            app = self._apps.get(self._hosts.get(origin, ""))
            if app is not None:
                context = self.by_id(app["tenantId"], app["id"])
                if context is not None:
                    return context
        return None
